pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/StateVectorLQubitDynamic.hpp

```diff
@@ -0,0 +1,387 @@
+// Copyright 2022 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <numeric>
+#include <utility>
+#include <vector>
+
+#include "BitUtil.hpp"
+#include "LinearAlgebra.hpp"
+#include "Util.hpp"
+
+#include "Error.hpp"
+
+#include <StateVectorLQubit.hpp>
+
+#include <iostream>
+
+namespace {
+using Pennylane::Util::AlignedAllocator;
+using Pennylane::Util::bestCPUMemoryModel;
+using Pennylane::Util::exp2;
+using Pennylane::Util::isPerfectPowerOf2;
+using Pennylane::Util::log2PerfectPower;
+using Pennylane::Util::ONE;
+using Pennylane::Util::squaredNorm;
+using Pennylane::Util::ZERO;
+} // namespace
+
+namespace Pennylane::LightningQubit {
+/**
+ * @brief State-vector dynamic class.
+ *
+ * This class allocates and deallocates qubits/wires dynamically,
+ * and defines all operations to manipulate the statevector data for
+ * quantum circuit simulation.
+ *
+ */
+template <class fp_t = double>
+class StateVectorLQubitDynamic
+    : public StateVectorLQubit<fp_t, StateVectorLQubitDynamic<fp_t>> {
```

Comment:
I wonder if we need this class to be downstreamed to Lightning! cc: @mlxd @dime10 @erick-xanadu 

Catalyst doesn't support the dynamic features of this class and Lightning simulators don't care about this dynamism for qubit allocations/releases. What do you think of removing this class and use `StaveVectorLQubitManaged` similar to LK and LGPU instead?

Reply:
Any comments on this? @mlxd @erick-xanadu

Reply:
I tried to remove it and noticed it would require some good effort for making it work with `StaveVectorLQubitManaged`. Since the tests have been migrated and seem to be working now, what do you think if we tackle this in a follow-up story? @maliasadi

Reply:
Sounds good @rauletorresc! Could you share the SC ticket here?

Reply:
https://app.shortcut.com/xanaduai/story/77689/replace-stavevectorlqubitdynamic-with-stavevectorlqubitmanaged I scheduled it for Q4 Iteration 4